Implement scrollbar-width CSS property in WebKit2

lukewarlow opened the pull request:
#### 7a7bb7a3e5e23541e6ca556920b8359cf9c7b785
<pre>
Implement scrollbar-width CSS property in WebKit2
<a href="https://bugs.webkit.org/show_bug.cgi?id=256999">https://bugs.webkit.org/show_bug.cgi?id=256999</a>

Reviewed by Simon Fraser.

This change implements the basics of the scrollbar-width CSS property.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-keywords-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::createScrollbar):
(WebCore::LocalFrameView::scrollbarWidthStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::Scrollbar):
(WebCore::Scrollbar::isHiddenByStyle const):
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::isHiddenByStyle const): Deleted.
* Source/WebCore/platform/ScrollbarTheme.h:
(WebCore::ScrollbarTheme::scrollbarThickness):
* Source/WebCore/platform/ScrollbarThemeComposite.cpp:
(WebCore::ScrollbarThemeComposite::minimumThumbLength):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::scrollbarThickness):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.mm:
(WebCore::ScrollbarThemeIOS::scrollbarThickness):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::scrollbarThickness):
(WebCore::ScrollbarThemeMac::backButtonRect):
(WebCore::ScrollbarThemeMac::forwardButtonRect):
(WebCore::ScrollbarThemeMac::trackRect):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.h:
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.cpp:
(WebCore::ScrollbarThemePlayStation::scrollbarThickness):
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.h:
* Source/WebCore/platform/win/ScrollbarThemeWin.cpp:
(WebCore::ScrollbarThemeWin::scrollbarThickness):
* Source/WebCore/platform/win/ScrollbarThemeWin.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
(WebCore::RenderLayerScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderScrollbar.h:
* Source/WebCore/rendering/RenderScrollbarTheme.h:

Canonical link: <a href="https://commits.webkit.org/264593@main">https://commits.webkit.org/264593@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2334d700df4cf0ecc424da09649af633a41e1355

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9670 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8122 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8214 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10987 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9249 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9790 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14918 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7673 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10821 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7930 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6451 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7244 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1935 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7671 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -538,8 +538,12 @@ Ref<Scrollbar> LocalFrameView::createScrollbar(ScrollbarOrientation orientation)
     if (frameRenderer && frameRenderer->style().hasPseudoStyle(PseudoId::Scrollbar))
         return RenderScrollbar::createCustomScrollbar(*this, orientation, nullptr, m_frame.ptr());
 
+    ScrollbarControlSize size = scrollbarWidth() == ScrollbarWidth::Thin
```
`auto scrollbarSize = ` (better than just "size" because size is normally a thing with height and width).

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -6278,6 +6282,15 @@ OverscrollBehavior LocalFrameView::verticalOverscrollBehavior()  const
+ScrollbarWidth LocalFrameView::scrollbarWidth()  const
+{
+    auto* document = m_frame->document();
+    auto scrollingObject = document && document->documentElement() ? document->documentElement()->renderer() : nullptr;
```
Is `documentElement` the right thing to check for the style? How about `scrollingElement`? Does `scrollbar-width` propagate from the body to the document like `overflow` does?

lukewarlow replied:
This reads to me like the answer is no it doesn't propagate and as such I think you're right it should be scrollingElement not document element.

https://drafts.csswg.org/css-scrollbars/#scrollbar-width:~:text=is%20not%20propagated%20to%20the%20viewport.

![image](https://github.com/WebKit/WebKit/assets/32498324/d330e4a5-eeee-4a28-a0a2-3893815d404e)

smfr replied:
Make sure there at WPT tests that test that.

nt1m replied:
@lukewarlow ^ can you please do this as a followup?

smfr commented on Source/WebCore/platform/ScrollbarTheme.h:
```diff
@@ -51,7 +51,7 @@ class ScrollbarTheme {
     virtual ScrollbarPart hitTest(Scrollbar&, const IntPoint&) { return NoPart; }
     
-    virtual int scrollbarThickness(ScrollbarControlSize = ScrollbarControlSize::Regular, ScrollbarExpansionState = ScrollbarExpansionState::Expanded) { return 0; }
+    virtual int scrollbarThickness(ScrollbarControlSize = ScrollbarControlSize::Regular, bool = false, ScrollbarExpansionState = ScrollbarExpansionState::Expanded) { return 0; }
```
The bool param needs a name (commented out, since the implementation is empty). Or it might be cleaner to just pass the ScrollbarWidth enum.

lukewarlow replied:
That makes sense I will replace it with the ScrollbarWidth enum. A follow up change could probably remove the ScrollbarControlSize enum entirely as ScrollbarWidth should be able to replace it entirely

lukewarlow replied:
@smfr should I make the bug to refactor ScrollbarControlSize away now or should I wait till this is merged and then do it?

smfr replied:
You can do it separately.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -867,7 +867,10 @@ Ref<Scrollbar> RenderLayerScrollableArea::createScrollbar(ScrollbarOrientation o
         widget = RenderScrollbar::createCustomScrollbar(*this, orientation, element);
     else {
-        widget = Scrollbar::createNativeScrollbar(*this, orientation, ScrollbarControlSize::Regular);
+        ScrollbarControlSize size = scrollbarWidth() == ScrollbarWidth::Thin
```
`auto scrollbarSize = `

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -141,6 +141,8 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea>, public CanMakeChec
     virtual OverscrollBehavior horizontalOverscrollBehavior() const { return OverscrollBehavior::Auto; }
     virtual OverscrollBehavior verticalOverscrollBehavior() const { return OverscrollBehavior::Auto; }
 
+    virtual ScrollbarWidth scrollbarWidth() const { return ScrollbarWidth::Auto; }
```
Slightly uneasy with the name (it'd expect it to return an `int`). Maybe `scrollbarWidthStyle()`?

smfr reviewed (changes requested):
Let's do an update passing ScrollbarWidth instead of the bool, and checking the style propagation question.

lukewarlow committed:
Implement scrollbar-width CSS property in WebKit2
https://bugs.webkit.org/show_bug.cgi?id=256999

Reviewed by Simon Fraser.

This change implements the basics of the scrollbar-width CSS property.

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-keywords-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-width-007-expected.txt.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/css/cssom-view/client-props-root-expected.txt.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::createScrollbar):
(WebCore::LocalFrameView::scrollbarWidthStyle const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::Scrollbar):
(WebCore::Scrollbar::isHiddenByStyle const):
* Source/WebCore/platform/Scrollbar.h:
(WebCore::Scrollbar::isHiddenByStyle const): Deleted.
* Source/WebCore/platform/ScrollbarTheme.h:
(WebCore::ScrollbarTheme::scrollbarThickness):
* Source/WebCore/platform/ScrollbarThemeComposite.cpp:
(WebCore::ScrollbarThemeComposite::minimumThumbLength):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.cpp:
(WebCore::ScrollbarThemeAdwaita::scrollbarThickness):
* Source/WebCore/platform/adwaita/ScrollbarThemeAdwaita.h:
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.cpp:
(WebCore::ScrollbarThemeGtk::scrollbarThickness):
* Source/WebCore/platform/gtk/ScrollbarThemeGtk.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.h:
* Source/WebCore/platform/ios/ScrollbarThemeIOS.mm:
(WebCore::ScrollbarThemeIOS::scrollbarThickness):
* Source/WebCore/platform/mac/ScrollbarThemeMac.h:
* Source/WebCore/platform/mac/ScrollbarThemeMac.mm:
(WebCore::ScrollbarThemeMac::scrollbarThickness):
(WebCore::ScrollbarThemeMac::backButtonRect):
(WebCore::ScrollbarThemeMac::forwardButtonRect):
(WebCore::ScrollbarThemeMac::trackRect):
* Source/WebCore/platform/mac/ScrollbarsControllerMac.mm:
(WebCore::ScrollbarsControllerMac::updateScrollerStyle):
* Source/WebCore/platform/mock/ScrollbarThemeMock.cpp:
(WebCore::ScrollbarThemeMock::scrollbarThickness):
* Source/WebCore/platform/mock/ScrollbarThemeMock.h:
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.cpp:
(WebCore::ScrollbarThemePlayStation::scrollbarThickness):
* Source/WebCore/platform/playstation/ScrollbarThemePlayStation.h:
* Source/WebCore/platform/win/ScrollbarThemeWin.cpp:
(WebCore::ScrollbarThemeWin::scrollbarThickness):
* Source/WebCore/platform/win/ScrollbarThemeWin.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
(WebCore::RenderLayerScrollableArea::scrollbarWidthStyle const):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebCore/rendering/RenderScrollbar.h:
* Source/WebCore/rendering/RenderScrollbarTheme.h:

Canonical link: https://commits.webkit.org/264593@main
